### PR TITLE
[JENKINS-41453] Fix fallout from fixing the overzealous migration of heads

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-41453](https://issues.jenkins-ci.org/browse/JENKINS-41453)

- The pull requests are missing semi-required information and we need to migrate the key fields to prevent a build storm

Downstream of https://github.com/jenkinsci/scm-api-plugin/pull/25

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 1]
```diff
@@ -825,7 +825,7 @@ public SCM build(SCMHead head, SCMRevision revision) {
                 setBrowser(scm, repoUrl);
             }
             return scm;
-        } else if (head instanceof PullRequestSCMHead) {
+        } else if (head instanceof PullRequestSCMHead && ((PullRequestSCMHead) head).getSourceRepo() != null) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: Does this go to the `GitSCM` fallback? Sounds wrong.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java [hunk 2]
```diff
@@ -114,4 +147,35 @@ public String getSourceBranch() {
     public String getSourceRepo() {
         return sourceRepo;
     }
+
+    @Restricted(NoExternalUse.class)
+    public static class Metadata {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: Make it `private`.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java [hunk 3]
```diff
@@ -43,6 +49,7 @@
     private final String sourceOwner;
     private final String sourceRepo;
     private final String sourceBranch;
+    private transient Metadata metadata;
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: Maybe I'm missing something but, where is this set?

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: Deserialising legacy implementation

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: i believe it is only theoretically positive end up here with the null... but if it does then we will fall through... 

But the git impl doesn't know how to handle pull request heads, so should return null itself... unless you happen to have called your branch PR-123 in which case the PR branch would be masked by the non PR branch with a name using the PR branch name pattern so you couldn't get here anyway.

If I think I follow what I just wrote... probably fine!

**Comment by svanoort** [comment]
🐝 as far as I understand it

**Pull request merged**